src/language_server_protocol.rs

```diff
@@ -2103,6 +2347,182 @@ impl LanguageClient {
+                            group: group.clone(),
+                            text: String::new(),
+                        });
+
+                        line_has_sym = true;
+                    }
+                }
+
+                if line_has_sym {
+                    // Try to coalesce adjacent lines to reduce the API calls
+                    match clears.last_mut() {
+                        Some(last) => {
+                            if last.line_end == line.line as u64 {
+                                last.line_end += 1;
+                            } else {
+                                clears.push(ClearNamespace {
+                                    line_start: line.line as u64,
+                                    line_end: line.line as u64 + 1,
+                                });
+                            }
+                        }
+                        None => {
+                            clears.push(ClearNamespace {
+                                line_start: line.line as u64,
+                                line_end: line.line as u64 + 1,
+                            });
+                        }
+                    }
+                }
+            }
+
+            let mut num_semantic_hls = 0;
+            let num_new_semantic_hls = highlights.len();
+
+            self.update(|state| {
+                state
+                    .semantic_highlights_syms
+                    .insert(languageId.clone(), params.lines);
+
+                state.vim.rpcclient.notify(
+                    "s:ApplySemanticHighlights",
+                    json!([buffer, ns_id, clears, highlights]),
+                )?;
+
+                let mut old_semantic_hls = state
+                    .semantic_highlights
+                    .insert(languageId.clone(), Vec::new())
+                    .unwrap_or_default();
+
+                // FIXME: These are most likely invalid due to new/deleted lines
+                //        shifting the highlighted regions from their original positions
+                //
+                //        Technically this would break incremental semantic highlighting
+                //        but it seems like no server implements this right now
+                //
+                //        If this is not cleared the array of highlights can grow very large.
+                old_semantic_hls.clear();
+
+                let mut existing_hls = old_semantic_hls.into_iter().peekable();
+                let mut new_hls = highlights.into_iter().peekable();
+                let semantic_hls = state.semantic_highlights.get_mut(&languageId).unwrap();
+
+                loop {
+                    match (existing_hls.peek(), new_hls.peek()) {
+                        (Some(existing_hl), Some(new_hl)) => {
+                            use std::cmp::Ordering;
+
+                            match existing_hl.line.cmp(&new_hl.line) {
+                                Ordering::Less => {
```

Comment:
Instead of comparing each highlight group, can we just delete all highlights on screen and add all newer visible highlights? Similarly to how diagnostics are highlighted. In my opinion, the implementation would be much simplier.

Reply:
The main issue is that the protocol supports incremental highlighting. So if that case is not handled then it may only apply the changed highlighting.

Currently `eclipse.jdt.ls` sends a single line of highlighting if the line itself was changed, if a line was added/removed the entire document's worth of highlighting is sent.

From this comment (https://github.com/prabirshrestha/vim-lsp/issues/633#issuecomment-573323937) it looks like `clangd 10` has similar behavior except it can also send half the document's worth of highlighting upon add/remove lines.

I'll continue looking at simplifying this logic a bit. Might try a region based approach rather than the current line by line one.